plugins/destination/s3/client/write.go

```diff
@@ -70,9 +72,73 @@ func (c *Client) createObject(ctx context.Context, table *schema.Table, objKey s
 			params.ServerSideEncryption = sseConfiguration.ServerSideEncryption
 		}
 
-		_, err = manager.NewUploader(c.s3Client, func(uploader *manager.Uploader) {
+		uploader := manager.NewUploader(c.s3Client, func(uploader *manager.Uploader) {
 			uploader.PartSize = *c.spec.PartSize
-		}).Upload(ctx, params)
+		})
+
+		// Retry configuration: 3 retries (4 attempts total) with exponential backoff
+		const maxAttempts = 4
+		const minBackoff = 100 * time.Millisecond
+		const maxBackoff = 3 * time.Second
+
+		// If the reader supports seeking, we can safely retry failed uploads by seeking back to the
+		// start before each attempt (so the uploader can re-read the full content).
+		// If it doesn't (NoRotate mode), fall back to a single attempt since the data can't be re-read.
+		if rs, ok := adjustedReader.(io.ReadSeeker); ok {
```

Comment:
In other places (SDK, plugins) we use https://github.com/hashicorp/go-retryablehttp